crates/ipc/src/bridge/buffer.rs

```diff
@@ -0,0 +1,242 @@
+}
+
+#[repr(C, align(64))]
+struct CachePaddedAtomicU32(AtomicU32);
+
+#[derive(Debug, thiserror::Error)]
+pub enum WriteError {
+    #[error("input too large (tried to write {0} bytes)")]
+    TooLarge(usize),
+
+    #[error("the buffer is full")]
+    Full,
+}
+
+impl BipBuffer {
+    const ALIGN: u32 = 16;
+
+    pub const fn new() -> Self {
+        let dangling = NonNull::dangling();
+        Self {
+            buf: unsafe { RelPtr::new(dangling, dangling.cast()) },
+            len: 0,
+            lock: CachePaddedAtomicU32::new(0),
+            w: CachePaddedAtomicU32::new(0),
+            r: CachePaddedAtomicU32::new(0),
+        }
+    }
+
+    pub unsafe fn init(&mut self, buf: NonNull<u8>, len: u32) {
+        // SAFETY: up to caller.
+        self.buf = unsafe { RelPtr::new(buf, NonNull::from_ref(self).cast()) };
+
+        // The end is not aligned to 16 bytes, so it can always fit a LEB but not a message.
+        self.len = u32::saturating_sub(len & Self::ALIGN.wrapping_neg(), 1);
+    }
+
+    /// # Safety
+    ///
+    /// Only one thread is permitted to read from the buffer at a time.
+    pub unsafe fn read<T, F: FnOnce(&mut [u8]) -> T>(&self, f: F) -> Option<T> {
+        let r = self.r.load(Ordering::Relaxed);
+        let w = self.w.load(Ordering::Acquire);
+
+        if r == w {
+            return None;
+        }
+
+        let buf = self.buf();
+        let read_start = unsafe { buf.add(r as usize) };
+
+        let mut cursor = read_start;
+
+        // Read the next message length.
+        // Note how it's unconditionally at the read position if the buffer is non-empty.
+        let len = unsafe { read_leb128(&mut cursor) };
+
+        let mut new_pos =
+            unsafe { u32::next_multiple_of(cursor.offset_from_unsigned(buf) as u32, Self::ALIGN) };
+
+        if new_pos + len + LEB128_CAP >= self.len {
+            // The message is at the start of the buffer.
+            new_pos = 0;
+        }
+
+        cursor = unsafe { buf.add(new_pos as usize) };
+        let bytes = unsafe { NonNull::slice_from_raw_parts(cursor, len as usize).as_mut() };
+
+        let value = f(bytes);
+
+        let new_r = new_pos + len;
+        self.r.store(new_r, Ordering::Release);
+
+        Some(value)
+    }
+
+    pub fn write(&self, bytes: &[u8]) -> Result<(), WriteError> {
+        let len = bytes.len();
+        let leb_len = size_of_leb128(len as u32);
+
+        let max_len = len + leb_len as usize + Self::ALIGN as usize - 1;
+        if max_len > self.len as usize / 2 {
+            // Don't try to write messages over half the length of the buffer.
+            return Err(WriteError::TooLarge(len));
+        }
+        let len = len as u32;
+
+        let r = self.r.load(Ordering::Acquire);
+
+        // Have to spin to make writes atomic.
+        struct LockGuard<'a>(&'a AtomicU32);
+
+        impl Drop for LockGuard<'_> {
+            fn drop(&mut self) {
+                self.0.store(0, Ordering::Relaxed);
```

Comment:
Shouldn't the lock be dropped with `Release` ordering? Without this there is no guarantee that writes done while holding the lock will be visible to concurrent writers. 

In particular this means that the `w` load at line 138 is not synchronized with the store at line 168, which is a race condition (well, if we weren't on x86 where every load/store is AcqRel).

```suggestion
                self.0.store(0, Ordering::Release);
```